serokell-style.md

```diff
@@ -338,9 +345,39 @@ data Address
     deriving (Show, Eq)
 ```
 
+Type classes in `deriving` section should be always surrounded by
+parentheses. Space between names is optional.
+
+_WARNING_: try to avoid aggressive autoderiving. Deriving instances can
+slowdown compilation
+(stated here: http://www.stephendiehl.com/posts/production.html)
+
+> Deriving instances of Read/Show/Data/Generic for largely recursive ADTs can
+> sometimes lead to quadratic memory behavior when the nesting gets deep.
+
+If you're using GHC-8.2.2 or higher you should use
+[`-XDerivingStrategies`](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#deriving-strategies)
+extension and specify the way you derive explicitly. Like this (see
+[this question](https://github.com/jaspervdj/stylish-haskell/issues/186) for example):
+
+```haskell
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE DerivingStrategies         #-}
+
+newtype SpecialId = SpecialId Int
+    deriving stock    (Eq, Ord, Show, Generic)
+    deriving newtype  Read
+    deriving anyclass (FromJSON, ToJSON)
+```
+
 ### Function declaration
 
-All functions must have type signatures.
+All top-level functions _must_ have type signatures.
+
+All functions inside `where` _must_ have type signatures.
```

Comment:
Nope.

Reply:
You may have strange errors which are hard to resolve if you don't specify types of functions inside `where`. Because type inference can be your enemy in that case. Though I'm okay with relaxing `must` to `should`.

Reply:
Should is good. Sometimes I work with pure arithmetics and everything is, say, `Integer` or a function on integers. It's stupid to require writing this.